pkg/syncutil/cachemanager/cachemanager.go

```diff
@@ -77,6 +228,139 @@ func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.
 	return nil
 }
 
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
 func (c *CacheManager) ReportSyncMetrics() {
 	c.syncMetricsCache.ReportSync()
 }
+
+func (c *CacheManager) syncGVKInstances(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := c.reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+var gvksFailingTolist *watch.Set
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	// stopChan is used to stop any list operations still in progress
+	stopChan := make(chan bool, 1)
+	gvksFailingTolist = watch.NewSet()
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(stopChan)
+			return
+		case <-c.backgroundManagementTicker.C:
+			c.mu.Lock()
+			c.wipeCacheIfNeeded(ctx)
+
+			// spin up new goroutines to relist if new gvks to relist are
+			// populated from makeUpdates.
+			if c.gvksToList.Size() != 0 {
+				// stop any goroutines that were relisting before
+				// as we may no longer be interested in those gvks
+				stopChan <- true
+
+				// also try to catch any gvks that are in the aggregator
+				// but are failing to list from a previous replay.
+				for _, gvk := range gvksFailingTolist.Items() {
+					if c.specifiedGVKs.IsPresent(gvk) {
+						c.gvksToList.Add(gvk)
+					}
+				}
+
+				// save all gvks that need relisting
+				gvksToRelistForLoop := c.gvksToList.Items()
+
+				// clean state
+				gvksFailingTolist = watch.NewSet()
+				c.gvksToList = watch.NewSet()
+
+				stopChan = make(chan bool, 1)
+
+				go c.replayLoop(ctx, gvksToRelistForLoop, stopChan)
+			}
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (c *CacheManager) replayLoop(ctx context.Context, gvksToRelist []schema.GroupVersionKind, stopChan <-chan bool) {
```

Comment:
naming nit:  Adding a programming construct like "Loop" doesn't tell us much.  That's like `replayConditionalLogic` or `replaySwitch`.  Let's go with `replay`, or if you want to communicate that it iterates over a set of data, we can say that with `replayGVKs`.